### ISSUE TITLE
import-url from a google bucket crashes with file not found

### ISSUE THREAD

**Issue by D34D10CK** [iterative/dvc#8026]
# Bug Report

`import-url` fails to import from bucket

## Description

When using `import-url` to import data from a bucket, the command fails with a file not found error. DVC manages to correctly download all files but crashes afterwards. 
The stack trace I get while running `dvc import-url gs://my-bucket data/raw -v` is the following (I had to anonymize the names of the files):
```
Traceback (most recent call last):
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/cli/__init__.py", line 185, in main
    ret = cmd.do_run()
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/cli/command.py", line 22, in do_run
    return self.run()
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/commands/imp_url.py", line 15, in run
    self.repo.imp_url(
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/repo/__init__.py", line 48, in wrapper
    return f(repo, *args, **kwargs)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/repo/scm_context.py", line 156, in run
    return method(repo, *args, **kw)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/repo/imp_url.py", line 83, in imp_url
    stage.run(jobs=jobs)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/funcy/decorators.py", line 45, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/stage/decorators.py", line 36, in rwlocked
    return call()
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/funcy/decorators.py", line 66, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/stage/__init__.py", line 549, in run
    self.save(allow_missing=allow_missing)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/stage/__init__.py", line 459, in save
    self.save_deps(allow_missing=allow_missing)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/stage/__init__.py", line 470, in save_deps
    dep.save()
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc/output.py", line 552, in save
    _, self.meta, obj = build(
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_data/build.py", line 240, in build
    meta, obj = _build_tree(
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_data/build.py", line 137, in _build_tree
    meta, obj = _build_file(
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_data/build.py", line 64, in _build_file
    meta, hash_info = hash_file(path, fs, name, state=state)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_data/hashfile/hash.py", line 178, in hash_file
    hash_value, info = _hash_file(path, fs, name, callback=cb)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_data/hashfile/hash.py", line 123, in _hash_file
    info = _adapt_info(fs.info(path), fs.protocol)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/dvc_objects/fs/base.py", line 346, in info
    return self.fs.info(path)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/fsspec/asyn.py", line 86, in wrapper
    return sync(self.loop, func, *args, **kwargs)
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/fsspec/asyn.py", line 66, in sync
    raise return_result
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/fsspec/asyn.py", line 26, in _runner
    result[0] = await coro
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/gcsfs/core.py", line 706, in _info
    if self._ls_from_cache(path):
  File "/home/user/.local/share/virtualenvs/projectlib/python3.10/site-packages/fsspec/spec.py", line 362, in _ls_from_cache
    raise FileNotFoundError(path)
FileNotFoundError: some_file_that_doesnt_exist.xlsx
```

If that can help, the file that DVC claims is missing used to exist in the bucket at some point in the past and some files inside sub folders in the bucket have the same name.
I tried downgrading to dvc 2.12.0 and the issue goes away.

### Reproduce

`dvc import-url gs://my-bucket data/raw`

### Expected

The command completes successfully.

### Environment information

**Output of `dvc doctor`:**

```console
$ dvc doctor
```

```
DVC version: 2.13.0 (pip)
---------------------------------
Platform: Python 3.10.4 on Linux-5.15.0-1010-gcp-x86_64-with-glibc2.35
Supports:
        gs (gcsfs = 2022.5.0),
        webhdfs (fsspec = 2022.5.0),
        http (aiohttp = 3.8.1, aiohttp-retry = 2.5.2),
        https (aiohttp = 3.8.1, aiohttp-retry = 2.5.2)
Cache types: hardlink, symlink
Cache directory: ext4 on /dev/sda1
Caches: local
Remotes: None
Workspace directory: ext4 on /dev/sda1
Repo: dvc, git
```

**Additional Information (if any):**

**Comment by dtrifiro** [comment]
Could not reproduce using the following reproduction script:
```bash
#!/bin/bash

rm -rf gh_8026
mkdir gh_8026
cd gh_8026

python -m venv .venv
source .venv/bin/activate

pip install 'dvc[gs]==2.13'
git init
dvc init
gsutil mb gs://testing-import-url

echo "test" >testfile
gsutil cp testfile gs://testing-import-url

dvc import-url gs://testing-import-url/testfile testfile_imported
```

Will try again later by adding/deleting files to try to reproduce your exact scenario

**Comment by D34D10CK** [comment]
i have not yet tried with your script, but i tried in a new bucket to add and delete files and I was able to reproduce the error.

**Comment by D34D10CK** [comment]
no, I created a new clean virtualenv everytime, I even tried doing it in a new vm in case it was related to some global config.

**Comment by dtrifiro** [comment]
I was able to reproduce the issue. Thank you for your help.

This is most likely related to the fact that you're importing the bucket root. A workaround, while we fix the issue in `dvc-data`, could be to move the data to a subdirectory of the bucket.

**Comment by dtrifiro** [comment]
Some more information:
For `gcsfs`, `fs.walk()` calls `ls()` -> `_list_objects()` (see https://github.com/fsspec/gcsfs/blob/f169fddbb0ace94418d806ba3a5c0706c7dd08db/gcsfs/core.py#L755-L767)

The first time `_list_objects`  is called, the dircache is empty, so the objects are listed and the dircache it is populated. On the second call, `_ls_from_cache` is used instead, which returns `testing-import-url/afolder/testfile`  instead of `testing-import-url/afolder`.

This results in `walk` returning `[('testing-import-url', , [testfile])` and in dvc trying to access a non-existing path

**Comment by dtrifiro** [comment]
A smaller reproduction script shows that a `fs.find` call corrupts the dircache by replacing a directory with all the files with the same prefix

```python
from gcsfs import GCSFileSystem

fs = GCSFileSystem(projectname="dvc-test-239404", bucket="testing-import-url")
print("Calling ls")
first_ls = fs.ls("testing-import-url")
print(f"{first_ls=}\n")

print("Calling find")
find = fs.find("testing-import-url")
print(f"{find=}\n")

print("Calling ls (again)")
second_ls = fs.ls("testing-import-url")
print(f"{second_ls=}\n")

print(f"new items: {set(second_ls)-set(first_ls)}")
print(f"removed items: {set(first_ls)-set(second_ls)}")
```
> Calling ls
> first_ls=['testing-import-url/afolder', 'testing-import-url/testfile']
> 
> Calling find
> find=['testing-import-url/afolder/subfolder/bar', 'testing-import-url/afolder/testfile', 'testing-import-url/testfile']
> 
> Calling ls (again)
> second_ls=['testing-import-url/afolder/subfolder/bar', 'testing-import-url/afolder/testfile', 'testing-import-url/testfile']
> 
> new items: {'testing-import-url/afolder/testfile', 'testing-import-url/afolder/subfolder/bar'}
> removed items: {'testing-import-url/afolder'}

**Comment by dtrifiro** [comment]
Waiting on https://github.com/fsspec/gcsfs/pull/488

**Comment by dtrifiro** [comment]
Temporarily patched `GCSFileSystem` in `dvc-gs` in https://github.com/iterative/dvc-gs/pull/9

**Comment by dtrifiro** [comment]
Fix was included `dvc-gs==2.19.1`, which is included starting with `dvc==2.22.0`
